Add message struct and tokio encoder/decoder

jonasnick opened the pull request.

real-or-random commented on src/messages.rs:
```diff
@@ -0,0 +1,299 @@
+        assert_eq!(codec.decode(&mut buf).unwrap().unwrap(), msg);
+        assert_eq!(codec.decode(&mut buf).unwrap().unwrap(), msg);
+        assert_eq!(buf.len(), 0);
+    }
```
A test for invalid public keys will be great.

jonasnick replied:
Fixed

real-or-random commented on src/messages.rs:
```diff
@@ -0,0 +1,299 @@
+}
+
+/// Codec implementing the Encoder and Decoder trait for
+/// Messages.
```
very short comment line, no linebreak necessary?

jonasnick replied:
Fixed

real-or-random commented on src/messages.rs:
```diff
@@ -0,0 +1,299 @@
+    ///
+    /// | Field      | Length | Description                                                        |
+    /// |------------|--------|--------------------------------------------------------------------|
+    /// | Signature  | 64     | Compressed ECDSA signature over the following public key using the |
```
This should be a Schnorr signature probably. I don't currently if elements allows to use keys generated for ECDSA in Schnorr signatures, I'll check

real-or-random commented on src/messages.rs:
```diff
@@ -0,0 +1,299 @@
+            return Ok(None);
+        }
+        let (magic, buf) = buf.split_at(4);
+        if !magic.starts_with(&[0x68, 0x1e, 0x58, 0x12]) {
```
Can we have the magic bytes as a constant? Also, a simple equality  check will be clearer I think

jonasnick replied:
fixed

real-or-random commented on src/messages.rs:
```diff
@@ -0,0 +1,299 @@
+/// | Field   | Length | Description                                                   |
+/// |---------|--------|---------------------------------------------------------------|
+/// | Variant | 1      | Determines which payload variant this is                      |
+/// | Variant | ...    | The actual variant                                            |
```
"Field" should not be "Variant" in second row

jonasnick replied:
fixed

real-or-random commented on src/messages.rs:
```diff
@@ -0,0 +1,299 @@
+    },
+}
+
+fn from_bytes_keyexchange<'a>(
```
This smells like we could be happy to have a Payload trait in the end (instead of an enum). I don't know, what do you think?

real-or-random commented on src/messages.rs:
```diff
@@ -0,0 +1,299 @@
+    /// |------------|--------|--------------------------------------------------------------------|
+    /// | Signature  | 64     | Compressed ECDSA signature over the following public key using the |
+    /// |            |        | long term  public key                                              |
+    /// | Public Key | 33     | Compressed ephemeral public key                                    |
```
The naming for public keys is difficult, we have many of them. But I have to talk to Andrew first before I know more.

real-or-random replied:
Okay, there won't be too many public keys. Still would be good to call that struct field `ke_pk` (for key exchange public key)

jonasnick replied:
Fixed

real-or-random commented on src/messages.rs:
```diff
@@ -0,0 +1,299 @@
+            }
+        }
+    }
+    fn to_bytes(&self, buf: &mut BytesMut, secp256k1: &Secp256k1) -> io::Result<()> {
```
Also here... looks like we will have a lot of boilerplate matches with the enum solution.

real-or-random commented:
Now the comments are ready. :)

jonasnick commented:
Added a commit to get rid of the ugly `from_bytes_keyexchange` by having each Payload variant store a struct corresponding to the message. Also renamed from_bytes to decode. Looks like less boilerplate to me now. Using a trait instead of a Payload enum doesn't seem to be the best option here. You'd have to use a trait objects with dynamic dispatch which has all sorts of [drawbacks](https://doc.rust-lang.org/book/first-edition/trait-objects.html#dynamic-dispatch) and would require additional boilerplate. We're also using enums in the functionary code.
